### DISCUSSION TITLE
OpenAI API Key

### DISCUSSION THREAD

**Discussion by MannyFay** [jackMort/ChatGPT.nvim#246] [Q&A]
Is there a problem if I set up the OpenAI API key with GPG with a passphrase?

I've tried to use it like in the plugin documentation:
```
The following configuration would use GPG to decrypt a local file containing the API key

local home = vim.fn.expand("$HOME")
require("chatgpt").setup({
    api_key_cmd = "gpg --decrypt " .. home .. "/secret.txt.gpg"
})
```
If I print it out in my terminal with `gpg --decrypt secret.txt.gpg` the GPG prompt asks for my passphrase. I pass it - the API key is correct.

But if I start Neovim, the message occors:  
[ChatGPT] [WARN 09:23:59] my-home-directory/.local/share/nvim/site/pack/packer/start/ChatGPT.nvim/lua/chatgpt/api.lua:163: Config 'api_key_cmd' did not return a value when executed

Is that because of the passphrase?

**Comment by KapJ1coH** [comment]
I had this problem. How did you encrypt the file?

**Reply by woohung** [reply]
Hi!
I encrypted it like this: `gpg -c secret.txt` there was an api string inside (without quotes). If i do `gpg --decrypt ~/secret.txt.gpg` in the console everything works correctly, a line with a token is returned

NVIM v0.9.5 (AstroNvim)
Build type: Release
LuaJIT 2.1.1692716794

```lua
return {
  "jackMort/ChatGPT.nvim",
    event = "VeryLazy",
    config = function()
      require("chatgpt").setup(
      {
        api_key_cmd = "gpg --decrypt ~/secret.txt.gpg"
      }
    )
    end,
    dependencies = {
      "MunifTanjim/nui.nvim",
      "nvim-lua/plenary.nvim",
      "folke/trouble.nvim",
      "nvim-telescope/telescope.nvim"
    }
}
```
**upd1**: I found an issue describing that `~` is no longer processed, so you need to use a construction like `local home = vim.fn.expand("$HOME")` and add type `.. home ..` to `api_key_cmd`.
But this also did not fix the problem.
**upd2**: I checked `:messages`  and found an error there: `[ChatGPT] [WARN  18:46:24] ~/.local/share/nvim/lazy/ChatGPT.nvim/lua/chatgpt/api.lua:197: Config 'api_key_cmd' did not return a value wh
en executed`
I found a thread in issue [287](https://github.com/jackMort/ChatGPT.nvim/issues/287), it is already closed, but the solution proposed there does not solve the problem. It is the same as in upd1.

My config look like this now:
```lua
return {
  "jackMort/ChatGPT.nvim",
  event = "VeryLazy",
  config = function()
    local home = vim.fn.expand("$HOME")
    require("chatgpt").setup({
      api_key_cmd = "gpg --decrypt " .. home .. "/secret.txt.gpg",
    })
  end,
  requires = {
    "MunifTanjim/nui.nvim",
    "nvim-lua/plenary.nvim",
    "folke/trouble.nvim",
    "nvim-telescope/telescope.nvim"
  }
}
```

**Reply by baggiponte** [reply]
How do you make gpg work? I get the same error but it never prompts me for the password.

**Reply by woohung** [reply]
Yep, same question.
Without a passphrase, the file cannot be encrypted via gpg, as I understand it.
`gpg --batch --passphrase '' -c secret.txt` doesn't work (with empty line)

I tried adding `api_key_cmd = "gpg --batch --passphrase 'devnet' --decrypt " .. home .. "/secret.txt.gpg"` doesn't work either. But this option works in the console, something like this `gpg --batch --passphrase 'devnet' --decrypt ~/secret.txt.gpg`

**Reply by johnnguyencodes** [reply]
I was encountering the same `'api_key_cmd' did not return a value when executed` issue but I managed to resolve it!  In my case, I needed to `gpg decrypt` the `secret.txt.gpg` file in terminal first, and then `api_key_cmd` will return the correct value in NeoVim.  

Steps to reproduce:
1. In `secret.txt`, I entered only the string value of the `api key`.  
2. After setting up the `api_key_cmd = "gpg --decrypt " .. home .. "/secret.txt.gpg"` as in @woohung's updated config, I went to my terminal and executed the `gpg --decrypt " ..home.. "/secret.txt.gpg"` command.  
3. The terminal prompted me to enter my password.  
4. After successfully entering the password, `gpg` correctly outputed the value in the terminal.  
5. Restart NeoVim and the `ChatGPT` plugin should now be able to use your secret api key.  

It seems like you do need to re-enter the password after some time.  When the `'api_key_cmd' did not return a value when executed` error appears again, I would go back to my terminal, run the same `gpg --decrypt` command to enter my password, and the error message will go away.

I wish the ChatGPT plugin would have a native way to prompt entering the `gpg decrypt` password inside NeoVim though, but  at least there is now a work around.
